File: domain/src/main/kotlin/com/kyovo/domain/model/booking/Booking.kt
package com.kyovo.domain.model.booking

import com.kyovo.domain.model.room.RoomId
import com.kyovo.domain.model.user.UserId

data class Booking(
    val id: BookingId,
    val roomId: RoomId,
    val userId: UserId,
    val startDate: BookingStartDate,
    val endDate: BookingEndDate,
    val numberOfPeople: BookingNumberOfPeople,
    val specialRequests: BookingSpecialRequests?,
    val statusInfo: BookingStatusInfo
)
{
    val status: BookingStatus get() = statusInfo.status

    fun transitionTo(
        target: BookingStatus,
        now: BookingStatusInfoDate,
        changedBy: UserId?,
        reason: BookingStatusReason?
    ): Booking?
    {
        if (!statusInfo.status.canTransitionTo(target)) return null
        return copy(statusInfo = BookingStatusInfo(target, now, changedBy, reason))
    }
}

File: frontend/src/domain/model/Booking.ts
export type BookingStatus = 'PENDING' | 'CONFIRMED' | 'CANCELED'

export type BookingStatusInfo = {
    status: BookingStatus
    since: string
    changedBy: string | null
    reason: string | null
}

export type Booking = {
    id: string
    roomId: string
    userId: string
    startDate: string
    endDate: string
    numberOfPeople: number
    specialRequests: string | null
    status: BookingStatus
    statusInfo: BookingStatusInfo
}

export type NewBooking = {
    roomId: string
    userId: string
    startDate: string
    endDate: string
    numberOfPeople: number
    specialRequests?: string
}
